src/HazelcastClient.ts

```diff
@@ -151,14 +147,16 @@ export default class HazelcastClient {
      */
     getDistributedObjects(): Promise<DistributedObject[]> {
         const clientMessage = ClientGetDistributedObjectsCodec.encodeRequest();
-        const toObjectFunc = this.getSerializationService().toObject.bind(this);
         const proxyManager = this.proxyManager;
-        return this.invocationService.invokeOnRandomTarget(clientMessage).then(function (resp): any {
-            const response = ClientGetDistributedObjectsCodec.decodeResponse(resp, toObjectFunc).response;
-            return response.map((objectInfo: { [key: string]: any }) => {
-                return proxyManager.getOrCreateProxy(objectInfo.value, objectInfo.key, false).value();
+        return this.invocationService.invokeOnRandomTarget(clientMessage)
+            .then((resp) => {
+                const response = ClientGetDistributedObjectsCodec.decodeResponse(resp).response;
+                return response.map((objectInfo) => {
+                    // TODO value throws if the returned promise from the getOrCreate is not fullfiled yet.
+                    //  This needs to be fixed. Also, we should create local instances instead of making remote calls.
+                    return proxyManager.getOrCreateProxy(objectInfo.name, objectInfo.serviceName, false).value();
```

Comment:
The `.value()` problem can be easily solved by using `Promise.all()` to wait for all `getOrCreateProxy()` calls to be resolved.

As for the "Also, we should create local instances instead of making remote calls" part, I see that Java client initializes all proxies (and re-initializes local ones) when processing this call. Personally, I find the return type of this method really weird and, as a user, I would expect to receive POJOs that describe DSes instead of proxy objects.

Reply:
I agree on the return type. Also, the Java client destroys the local instances of the destroyed proxies as a side effect. Added a TODO about this missing effect on this client. What do you think about it ?

Reply:
I think we should try to have the same behavior in the initial implementation. But first we need to understand what was the reason.

@sancar do you know why Java client destroys the local instances and then recreates them?
https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/client/impl/clientside/HazelcastClientInstanceImpl.java#L609

Reply:
@puzpuzpuz getDistributedObjects call cleans up dangling references on the client side.
The ones that we delete are already destroyed on the cluster, and should be deleted on the client side as well.

Reply:
So, removal of dangling proxies is the only reason for this side effect. It sounds like we should do the same in Node.js client.

Reply:
I added the removal logic to this method and related tests